Filebeat 6.2.1 in Kubernetes Segfaults

I was tailing the log of Filebeat because I saw that my Filebeat daemonset pods kept restarting.

This is the traceback:
```
2018-02-13T16:32:02.764Z	INFO	log/harvester.go:216	Harvester started for file: /var/lib/docker/containers/ab013b90ca95cb1a0d848ce537844044d7a05de434c1dd44199edcf36e99cb51/ab013b90ca95cb1a0d848ce537844044d7a05de434c1dd44199edcf36e99cb51-json.log
2018-02-13T16:32:02.764Z	ERROR	log/harvester.go:243	Read line error: decoding docker JSON: json: cannot unmarshal number into Go value of type reader.dockerLog; File: %!(EXTRA string=/var/lib/docker/containers/ab013b90ca95cb1a0d848ce537844044d7a05de434c1dd44199edcf36e99cb51/ab013b90ca95cb1a0d848ce537844044d7a05de434c1dd44199edcf36e99cb51-json.log)
2018-02-13T16:32:07.990Z	INFO	[monitoring]	log/log.go:124	Non-zero metrics in the last 30s	{"monitoring": {"metrics": {"beat":{"cpu":{"system":{"ticks":21760,"time":21760},"total":{"ticks":187170,"time":187172,"value":187170},"user":{"ticks":165410,"time":165412}},"info":{"ephemeral_id":"a0c1894e-0350-4d21-9f37-2df11e200e28","uptime":{"ms":11221079}},"memstats":{"gc_next":10567776,"memory_alloc":6966536,"memory_total":14985033360}},"filebeat":{"events":{"active":-6,"added":84,"done":90},"harvester":{"closed":3,"open_files":4,"running":4,"started":3}},"libbeat":{"config":{"module":{"running":1}},"output":{"events":{"acked":84,"batches":14,"total":84},"read":{"bytes":84},"write":{"bytes":13896}},"pipeline":{"clients":2,"events":{"active":1,"filtered":6,"published":78,"total":84},"queue":{"acked":84}}},"registrar":{"states":{"current":38,"update":90},"writes":18},"system":{"load":{"1":0.07,"15":1.5,"5":0.89,"norm":{"1":0.0088,"15":0.1875,"5":0.1113}}}}}}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x158 pc=0x13dd6fa]

goroutine 49 [running]:
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*ContainerIndexer).GetIndexes(0xc420428080, 0x0, 0xc42041d710, 0xc420436960, 0xc420436960)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/indexers.go:161 +0x3a
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*Indexers).GetIndexes(0xc42041d710, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/indexers.go:76 +0x230
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).removePod(0xc4201e2070, 0x0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:188 +0x3d
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).(github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.removePod)-fm(0x0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:159 +0x34
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.processEvent(0xc420469ed0, 0xc4201154d0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:175 +0x78
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).worker(0xc4201e2070)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:159 +0x26f
created by github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.newKubernetesAnnotator
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:113 +0x9f7
```

For confirmed bugs, please report:
- Version: 

`docker.elastic.co/beats/filebeat:6.2.1`

- Operating System: 

Kubernetes 1.9.1 on amd64

- Steps to Reproduce: 

Not sure

Yes actually, I got 2 more segafaults on exactly the node that had the "malformed" logs. This is on our staging cluster:
```
$ kubectl -n kube-system get pod -l k8s-app=filebeat
NAME             READY     STATUS    RESTARTS   AGE
filebeat-84fqs   1/1       Running   0          18h
filebeat-r8fq6   1/1       Running   1          18h
filebeat-rmk9w   1/1       Running   3          18h
filebeat-rxlp7   1/1       Running   0          18h
filebeat-tl76w   1/1       Running   0          18h
filebeat-z7rlh   1/1       Running   0          18h
```

The log of the crash can be reached using this:
```

$ kubectl -n kube-system logs filebeat-rmk9w --previous --tail 25
2018-02-14T04:30:19.307Z	INFO	log/harvester.go:216	Harvester started for file: /var/lib/docker/containers/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847-json.log
2018-02-14T04:30:19.321Z	ERROR	log/harvester.go:243	Read line error: decoding docker JSON: invalid character '\x00' looking for beginning of value; File: %!(EXTRA string=/var/lib/docker/containers/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847-json.log)
2018-02-14T04:30:29.311Z	INFO	log/harvester.go:216	Harvester started for file: /var/lib/docker/containers/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847-json.log
2018-02-14T04:30:29.324Z	ERROR	log/harvester.go:243	Read line error: decoding docker JSON: invalid character '\x00' looking for beginning of value; File: %!(EXTRA string=/var/lib/docker/containers/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847-json.log)
2018-02-14T04:30:39.314Z	INFO	log/harvester.go:216	Harvester started for file: /var/lib/docker/containers/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847-json.log
2018-02-14T04:30:39.328Z	ERROR	log/harvester.go:243	Read line error: decoding docker JSON: invalid character '\x00' looking for beginning of value; File: %!(EXTRA string=/var/lib/docker/containers/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847/c748a30c0c28fb6de5b53bffc1a8226a87fdca60d4df83640d889c2abe519847-json.log)
2018-02-14T04:30:47.889Z	INFO	[monitoring]	log/log.go:124	Non-zero metrics in the last 30s	{"monitoring": {"metrics": {"beat":{"cpu":{"system":{"ticks":2900,"time":2900},"total":{"ticks":21040,"time":21044,"value":21040},"user":{"ticks":18140,"time":18144}},"info":{"ephemeral_id":"2de41ee7-0730-4ed4-9360-e7f70241b11f","uptime":{"ms":2880248}},"memstats":{"gc_next":7829696,"memory_alloc":5643288,"memory_total":4368108992}},"filebeat":{"events":{"added":109,"done":109},"harvester":{"closed":3,"open_files":4,"running":4,"started":3}},"libbeat":{"config":{"module":{"running":1}},"output":{"events":{"acked":103,"batches":11,"total":103},"read":{"bytes":66},"write":{"bytes":14609}},"pipeline":{"clients":2,"events":{"active":0,"filtered":6,"published":103,"total":109},"queue":{"acked":103}}},"registrar":{"states":{"current":37,"update":109},"writes":13},"system":{"load":{"1":0.21,"15":0.44,"5":0.42,"norm":{"1":0.0263,"15":0.055,"5":0.0525}}}}}}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x158 pc=0x13dd6fa]

goroutine 28 [running]:
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*ContainerIndexer).GetIndexes(0xc4203f2a60, 0x0, 0xc4203f11d0, 0xc420158180, 0xc420158180)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/indexers.go:161 +0x3a
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*Indexers).GetIndexes(0xc4203f11d0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/indexers.go:76 +0x230
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).removePod(0xc4200a0000, 0x0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:188 +0x3d
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).(github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.removePod)-fm(0x0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:159 +0x34
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.processEvent(0xc4202e7ed0, 0xc4204bb920)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:175 +0x78
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).worker(0xc4200a0000)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:159 +0x26f
created by github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.newKubernetesAnnotator
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:113 +0x9f7
```

We do not have networking issues as far as I'm aware ...

And this is from the other node that had 1 restart:
```
kubectl -n kube-system logs filebeat-r8fq6 --previous --tail 25
2018-02-14T10:01:58.866Z	INFO	kubernetes/watcher.go:133	kubernetes: Watching API for pod events
2018-02-14T10:02:07.856Z	INFO	[monitoring]	log/log.go:124	Non-zero metrics in the last 30s	{"monitoring": {"metrics": {"beat":{"cpu":{"system":{"ticks":4460,"time":4468},"total":{"ticks":33970,"time":33980,"value":33970},"user":{"ticks":29510,"time":29512}},"info":{"ephemeral_id":"483fe03d-5e8d-4b07-ab05-bf62c71fb679","uptime":{"ms":59070199}},"memstats":{"gc_next":29848208,"memory_alloc":15164640,"memory_total":453574040}},"filebeat":{"harvester":{"open_files":5,"running":19}},"libbeat":{"config":{"module":{"running":1}},"pipeline":{"clients":2,"events":{"active":4117}}},"registrar":{"states":{"current":25}},"system":{"load":{"1":0.27,"15":0.19,"5":0.31,"norm":{"1":0.0338,"15":0.0238,"5":0.0388}}}}}}
2018-02-14T10:02:19.695Z	ERROR	pipeline/output.go:74	Failed to connect: dial tcp 10.106.82.87:12202: getsockopt: connection timed out
2018-02-14T10:02:37.855Z	INFO	[monitoring]	log/log.go:124	Non-zero metrics in the last 30s	{"monitoring": {"metrics": {"beat":{"cpu":{"system":{"ticks":4460,"time":4468},"total":{"ticks":33970,"time":33984,"value":33970},"user":{"ticks":29510,"time":29516}},"info":{"ephemeral_id":"483fe03d-5e8d-4b07-ab05-bf62c71fb679","uptime":{"ms":59100198}},"memstats":{"gc_next":29848208,"memory_alloc":15379400,"memory_total":453788800}},"filebeat":{"harvester":{"open_files":5,"running":19}},"libbeat":{"config":{"module":{"running":1}},"pipeline":{"clients":2,"events":{"active":4117,"retry":2048}}},"registrar":{"states":{"current":25}},"system":{"load":{"1":0.23,"15":0.19,"5":0.29,"norm":{"1":0.0288,"15":0.0238,"5":0.0363}}}}}}
2018-02-14T10:03:07.855Z	INFO	[monitoring]	log/log.go:124	Non-zero metrics in the last 30s	{"monitoring": {"metrics": {"beat":{"cpu":{"system":{"ticks":4470,"time":4472},"total":{"ticks":33980,"time":33988,"value":33980},"user":{"ticks":29510,"time":29516}},"info":{"ephemeral_id":"483fe03d-5e8d-4b07-ab05-bf62c71fb679","uptime":{"ms":59130198}},"memstats":{"gc_next":29848208,"memory_alloc":15563176,"memory_total":453972576}},"filebeat":{"harvester":{"open_files":5,"running":19}},"libbeat":{"config":{"module":{"running":1}},"pipeline":{"clients":2,"events":{"active":4117}}},"registrar":{"states":{"current":25}},"system":{"load":{"1":0.14,"15":0.18,"5":0.26,"norm":{"1":0.0175,"15":0.0225,"5":0.0325}}}}}}
2018-02-14T10:03:26.711Z	ERROR	pipeline/output.go:74	Failed to connect: dial tcp 10.106.82.87:12202: getsockopt: connection timed out
2018-02-14T10:03:37.856Z	INFO	[monitoring]	log/log.go:124	Non-zero metrics in the last 30s	{"monitoring": {"metrics": {"beat":{"cpu":{"system":{"ticks":4470,"time":4476},"total":{"ticks":34030,"time":34044,"value":34030},"user":{"ticks":29560,"time":29568}},"info":{"ephemeral_id":"483fe03d-5e8d-4b07-ab05-bf62c71fb679","uptime":{"ms":59160196}},"memstats":{"gc_next":29845296,"memory_alloc":14922760,"memory_total":454176392}},"filebeat":{"harvester":{"open_files":5,"running":19}},"libbeat":{"config":{"module":{"running":1}},"pipeline":{"clients":2,"events":{"active":4117,"retry":2048}}},"registrar":{"states":{"current":25}},"system":{"load":{"1":0.08,"15":0.17,"5":0.24,"norm":{"1":0.01,"15":0.0213,"5":0.03}}}}}}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x158 pc=0x13dd6fa]

goroutine 72 [running]:
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*ContainerIndexer).GetIndexes(0xc420499f80, 0x0, 0xc4204c30e0, 0xc420060cc0, 0xc420060c00)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/indexers.go:161 +0x3a
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*Indexers).GetIndexes(0xc4204c30e0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/indexers.go:76 +0x230
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).removePod(0xc4200984d0, 0x0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:188 +0x3d
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).(github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.removePod)-fm(0x0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:159 +0x34
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.processEvent(0xc42003eed0, 0xc420e9d1d0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:175 +0x78
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).worker(0xc4200984d0)
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:159 +0x26f
created by github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.newKubernetesAnnotator
	/go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:113 +0x9f7
```

I'm also experiencing same issue.

I'm using `docker.elastic.co/beats/filebeat:6.2.2` with kubernetes `v1.9.2`.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x158 pc=0x13dd8ba]

goroutine 96 [running]:
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*ContainerIndexer).GetIndexes(0xc42042bfc0, 0x0, 0xc420490bd0, 0xc420483680, 0xc420483620)
        /go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/indexers.go:161 +0x3a
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*Indexers).GetIndexes(0xc420490bd0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/indexers.go:76 +0x230
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).removePod(0xc4204bc0e0, 0x0)
        /go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:188 +0x3d
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).(github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.removePod)-fm(0x0)
        /go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:159 +0x34
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.processEvent(0xc4203e9ed0, 0xc42031d9b0)
        /go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:175 +0x78
github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.(*kubernetesAnnotator).worker(0xc4204bc0e0)
        /go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:159 +0x26f
created by github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata.newKubernetesAnnotator
        /go/src/github.com/elastic/beats/libbeat/processors/add_kubernetes_metadata/kubernetes.go:113 +0x9f7
```

--- a/CHANGELOG.asciidoc
+++ b/CHANGELOG.asciidoc
@@ -32,6 +32,8 @@ https://github.com/elastic/beats/compare/v6.2.2...6.2[Check the HEAD diff]
 
 *Affecting all Beats*
 
+- Avoid panic errors when processing nil Pod events in add_kubernetes_metadata. {issue}6372[6372]
+
 *Auditbeat*
 
 *Filebeat*

--- a/libbeat/processors/add_kubernetes_metadata/kubernetes.go
+++ b/libbeat/processors/add_kubernetes_metadata/kubernetes.go
@@ -168,8 +168,8 @@ func (k *kubernetesAnnotator) worker() {
 // Run pod actions while handling errors
 func processEvent(f func(pod *kubernetes.Pod), event bus.Event) {
 	pod, ok := event["pod"].(*kubernetes.Pod)
-	if !ok {
-		logp.Err("Couldn't get a pod from watcher event")
+	if !ok || pod == nil {
+		logp.Err("Couldn't get a pod from watcher event: %v", event)
 		return
 	}
 	f(pod)

